### ISSUE TITLE
How to set up sound device on PocketC.H.I.P. ?

### ISSUE THREAD

**Issue by dauuricus** [schismtracker/schismtracker#192]
I`ve compiled Schismtracker on PocketC.H.I.P. .

It is pretty good resolution for such a small computer screen.It is wonderfull.
But at defalt config setting in Schismtracker, PocketC.H.I.P. couldn't hold Schismtracker.

https://imgur.com/DkajnHf

"Couldn't open audio device: Device or resource busy"

ok,I need to edit config file...but How ?
https://imgur.com/LjgerzC

My pocketC.H.I.P.

```
chip@chip:~$ aplay --list-devices
**** List of PLAYBACK Hardware Devices ****
card 0: sun4icodec [sun4i-codec], device 0: CDC PCM Codec-0 []
  Subdevices: 0/1
  Subdevice #0: subdevice #0
```

```
chip@chip:~$ aplay -L
null
    Discard all samples (playback) or generate zero samples (capture)
pulse
    PulseAudio Sound Server
equal
default
dsp0
dmixer
sysdefault:CARD=sun4icodec
    sun4i-codec, 
    Default Audio Device
dmix:CARD=sun4icodec,DEV=0
    sun4i-codec, 
    Direct sample mixing device
dsnoop:CARD=sun4icodec,DEV=0
    sun4i-codec, 
    Direct sample snooping device
hw:CARD=sun4icodec,DEV=0
    sun4i-codec, 
    Direct hardware device without any conversions
plughw:CARD=sun4icodec,DEV=0
    sun4i-codec, 
    Hardware device with all software conversions
```

in .asoundrc

```

pcm.!default {
        type asym
        playback.pcm
        {
                type plug
                slave.pcm "dmixer"
        }
        capture.pcm
        {
                type plug
                slave.pcm "hw:0,0"
        }
}
pcm.dsp0 {
        type plug
        slave.pcm "dmixer"
}
pcm.dmixer {
        type dmix
        ipc_key 1024
        slave {
                pcm "hw:0,0"
                period_time 0
                period_size 1024
                buffer_size 8192
                rate 44100
        }
        bindings {
                0 0
                1 1
        }
}
ctl.mixer0 {
        type hw
        card 0
}

```
How to set audio device in config or is there any another way to set up audio device on  PocketC.H.I.P. ?

**Comment by dauuricus** [comment]
Although I do not know, when I try in this format, the sound comes to be played back.

https://imgur.com/NjOSuh6

```
[audio]
driver=alsa:default
```
https://imgur.com/hJ8mtbn

https://imgur.com/JvrJ5KW
https://imgur.com/S0Fj3ye
Using driver 'alsa'

**Comment by dauuricus** [comment]
I do not know how to write the value to pass to the SDL driver, but at this moment the sound was played on the schismtracker.

If there is a document that describes this in more detail, I would like to refer to it.

**Comment by jangler** [comment]
According to [docs/configuration.md#audio-output](https://github.com/schismtracker/schismtracker/blob/master/docs/configuration.md#audio-output), the `driver` option in the config file is parsed identically to the command-line `--audio-driver` option, which is documented in the man page:
```
       -a, --audio-driver=DRIVER[:DEVICE]
              Audio device configuration. driver is the SDL driver to use, e.g.  alsa (ALSA), dma or dsp (OSS); device is the name of the device itself, for example hw:2 or /dev/dsp1.
```
Does that answer your question?

**Comment by dauuricus** [comment]
Thank you, jangler !
I thought that nobody would give me an opinion on this matter.
Yes, I read the page about that configuration many times but I can't understand how to find correct working device in pocketCHIP and how to discribe about second argument.
I refer to the page of SDL,
https://wiki.libsdl.org/FAQUsingSDL

I predicted that the first argument might be "ALSA / alsa" or "PulseAudio / pulseaudio", but if you try to make the second argument "dmix" and start schism, schism will Because there is no driver, shcism automatically switches to the default device and reports that the device is busy.

Spent several hours with such a config rewrite and boot try-and-error :(

I did not know what that default audio device is and why schism can't interrupt that device, so I tested it with a milkytracker with pocketCHIP.

Because milkytracker requires minimum resolution of 640 * 480, it is difficult to use with a pocketCHIP screen resolution of 420 * 272, but when I try to read and play a .xm file with milkytracker, the sound comes out with headphones .

When I checked the milkytracker running on pocketCHIP at the terminal, SDL actually used "alsa".
https://imgur.com/wCBVviZ
Now I can predict that the first argument is "alsa". For the second argument, the candidate was "pcm", but trying it out was useless. But this is clearly moving forward, previously with the startup screen of schism,
"Using dummy on sdl"
It was
"Using alsa on sdl"
It seemed to have changed.

The second argument is still unknown, so I tested it with a likely value, and if it is "default", then it sounds safe to hear.

In other words, this is a state of testing with a brute-force attack rather than solving it.

So, in this case, your question, "Does that answer your question?", Is the answer "No".
I still do not know the correct answer logically about SDL audio driver.
I would like to refer to a more detailed description of this.

**Comment by dauuricus** [comment]
Here are sound driver in my PocketC.H.I.P. system.

```
chip@chip:~$ /sbin/lsmod | grep "snd"
snd_seq_dummy           1141  0 
snd_seq                49443  1 snd_seq_dummy
snd_seq_device          3045  1 snd_seq
snd_soc_core          116548  1 sun4i_codec
snd_pcm_dmaengine       2943  1 snd_soc_core
snd_pcm                68852  3 snd_soc_core,snd_pcm_dmaengine
snd_timer              17776  3 snd_pcm,snd_seq
snd                    47207  7 snd_soc_core,snd_timer,snd_pcm,snd_seq,snd_seq_device
soundcore               4991  1 snd
ac97_bus                1300  1 snd_soc_core
```
I checked sound device file "/dev/dsp" "/dev/audio" "snd_pcm" putting on second argument.
These three values were NG as an argument value.
https://imgur.com/86feFq4
https://imgur.com/Aw4F3HB
https://imgur.com/q9Id6Ti

So, what is the "default" of "alsa: default" referring to?

**Comment by jangler** [comment]
This is more of an ALSA question than a Schism question, but try `aplay -l`. I get a list of items that look like:
```
card 0: PCH [HDA Intel PCH], device 0: ALC892 Analog [ALC892 Analog]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: NVidia [HDA NVidia], device 3: HDMI 0 [HDMI 0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```
I believe that in this case `hw:0,0` corresponds to the first item (`card 0, device 0`) and `hw:1,3` corresponds to the second (`card 1, device 3`).

**Comment by dauuricus** [comment]
@jangler thank you.Yes,I think this is my knowledge problem :)
But the case of pocketchip or another small screen computer it is best or good choice shismtracker.
But at then a few user will fall in same audio out problem.
There are very few docs on archive on net about shismtracker. 
And I searched this kind of solution info, but it was nothing.
So I think it should be archive here.

**Comment by dauuricus** [comment]
@jangler it's mean in this case  "defalt" = hw 0,0 ?

And in this case "driver=alsa:hw:0" = "driver=alsa:default" ? Or "driver=alsa:hw:0,0" ?

https://imgur.com/CiSrYy1

```
chip@chip:~$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: sun4icodec [sun4i-codec], device 0: CDC PCM Codec-0 []
  Subdevices: 0/1
  Subdevice #0: subdevice #0
```

**Comment by dauuricus** [comment]
I tested 
case "driver=alsa:hw:0" 
https://imgur.com/ZSacU2s
It's not seems to be "defalt" device...
https://imgur.com/kohLQ1K

case "driver=alsa:hw:0,0"
https://imgur.com/n53FpWM
It's not seems to be "defalt" device too
https://imgur.com/6wmHW9U

Or is there another way of writing?
"driver=alsa:0"
https://imgur.com/ZUCQ3Kp

"driver=alsa:default"
https://imgur.com/i53HvDh

If I ask in another way, there is no documentation or info on what schism recognizes as "default" if you do not specify "driver =" in "config" in ".schism"?
When "Device init failed", I want to know which "defalt device" is retried and is "busy".

**Comment by dauuricus** [comment]
any way
"driver=alsa:defalt"
and
"driver=alsa:dsp0" / "driver=alsa:dmixer"
https://imgur.com/16cWgef
both works.

"driver=alsa:hw:0,0"
is not work.

```
ls -la /dev/snd
```
https://imgur.com/uj0blGl

**Comment by dauuricus** [comment]
What I wanted to know was written here.
https://github.com/schismtracker/schismtracker/blob/92e63a2ce9f84a0daabb8ff011067f0c9d97e658/schism/audio_playback.c
after 1231line
